lib/authoritex/geonames.ex

```diff
@@ -0,0 +1,157 @@
+
+  @impl Authoritex
+  def fetch(id) do
+    @http_uri_base <> geoname_id = id
+
+    case HTTPoison.get(
+           "http://api.geonames.org/getJSON",
+           [{"User-Agent", "Authoritex"}],
+           params: [
+             geonameId: geoname_id,
+             username: @username
+           ]
+         ) do
+      {:ok, %{body: response, status_code: 200}} ->
+        parse_fetch_result(response)
+
+      {:ok, %{body: response, status_code: status_code}} ->
+        {:error, parse_geonames_error(response, status_code)}
+
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  @impl Authoritex
+  def search(query, max_results \\ 30) do
+    case HTTPoison.get(
+           "http://api.geonames.org/searchJSON",
+           [{"User-Agent", "Authoritex"}],
+           params: [
+             q: query,
+             username: @username,
+             maxRows: max_results
+           ]
+         ) do
+      {:ok, %{body: response, status_code: 200}} ->
+        {:ok, parse_search_result(response)}
+
+      {:ok, %{body: response, status_code: status_code}} ->
+        {:error, parse_geonames_error(response, status_code)}
+
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  defp parse_search_result(response) do
+    response
+    |> Jason.decode!()
+    |> Map.get("geonames")
+    |> Enum.map(fn result ->
+      %{
+        id: @http_uri_base <> to_string(result["geonameId"]),
+        label: result["name"],
+        hint: parse_hint(result["adminName1"], result["countryName"])
+      }
+    end)
+  end
+
+  defp parse_fetch_result(%{"status" => %{"message" => message, "value" => error_code}}) do
+    {:error, "#{error_description(to_string(error_code))} (#{to_string(error_code)}). #{message}"}
+  end
+
+  defp parse_fetch_result(%{
+         "geonameId" => geoname_id,
+         "name" => name,
+         "adminName1" => adminName1,
+         "countryName" => countryName
+       }) do
+    hint = parse_hint(adminName1, countryName)
+
+    {:ok,
+     Enum.into(
+       [
+         id: @http_uri_base <> to_string(geoname_id),
+         label: name,
+         hint: hint,
+         qualified_label: Enum.join(Enum.filter([name, hint], & &1), ", ")
+       ],
+       %{}
+     )}
+  end
+
+  defp parse_fetch_result(response) do
+    case Jason.decode(response) do
+      {:ok, response} ->
+        parse_fetch_result(response)
+
+      {:error, error} ->
+        {:error, {:bad_response, error}}
+    end
+  end
+
+  defp parse_geonames_error(response, status_code) do
+    case Jason.decode(response) do
+      {:ok, %{"status" => %{"value" => 11}}} ->
+        status_code
+
+      {:ok, %{"status" => %{"message" => message, "value" => error_code}}} ->
+        "Status #{status_code}: #{error_description(to_string(error_code))} (#{
+          to_string(error_code)
+        }). #{message}"
+
+      {:error, error} ->
+        {:bad_response, error}
+    end
+  end
+
+  defp parse_hint(admin_name, country_name) do
```

Comment:
Geonames hinting/expansion is is one of the things I think QA does wrong. It always uses `adminName1` and `countryName` no matter what, so you end up with results like `Illinois, Illinois, United States` and `France, , France`. This implementation fixes the `, ,` issue, but still includes redundant subdivisions. I'd like to suggest hinting based on the result's `fcode`:

| fcode  | meaning                                                  | hint                         | examples                                                                                                     | note |
|--------|----------------------------------------------------------|------------------------------|--------------------------------------------------------------------------------------------------------------|------|
| `PCLI` | country                                                  | `nil`                        | `{label: "France", hint: nil}`<br/> `{label: "United States", hint: nil}`                                    |      |
| `RGN`  | region                                                   | `countryName` (if not blank) | `{label: "Midwest", hint: "United States"}`<br/> `{label: "Western Europe", hint: nil}`                      |already works this way|
| `ADM1` | state, province, other top-level administrative division | `countryName`                | `{label: "Illinois", hint: "United States"}`<br/> `{label: "Île-de-France", hint: "France"}`                 | |
| other  | anything else                                            | `adminName1`, `countryName`  | `{label: "Chicago", hint: "Illinois, United States"}`<br/> `{label: "Paris", hint: "Île-de-France, France"}` |already works this way|

This will lead to results for city-based landmarks kind of weirdly omitting their city names (e.g., `{label: "Eiffel Tower", hint: "Île-de-France, France"}` and `{label: "Wrigley Field", hint: "Illinois, United States"}`), but that can't easily be helped.